### PR TITLE
domain,infoschema: fix 'show tables' after snapshot read

### LINKED ISSUE #55835: `show tables` get wrong result after restart tidb and a snapshot read
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. use infoschema v2

```
set @@global.tidb_schema_cache_size = '512MB';
```

2. create table t and drop it, record the timestamp when t is available

```
use test;
create table t (id int);
select now();  ## get  '2024-09-04 09:48:54'
drop table t;
```

3. restart tidb

4. after a snapshot read, `show tables` goes wrong.

```
mysql> show tables;
Empty set (0.00 sec)

mysql> select * from t as of timestamp '2024-09-04 09:48:54';
Empty set (0.00 sec)

mysql> show tables;
+----------------+
| Tables_in_test |
+----------------+
| t              |
+----------------+
1 row in set (0.00 sec)
```

### 2. What did you expect to see? (Required)

```
mysql> show tables;
Empty set (0.00 sec)

mysql> show full tables;
Empty set (0.00 sec)

mysql> select * from information_schema.tables where table_schema = 't';
Empty set (0.00 sec)
```

### 3. What did you see instead (Required)

```
mysql> show tables;
+----------------+
| Tables_in_test |
+----------------+
| t              |
+----------------+
1 row in set (0.00 sec)

mysql> show full tables;
Empty set (0.00 sec)

mysql> select * from information_schema.tables where table_schema = 't';
Empty set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #55835

Problem Summary:

### What changed and how does it work?

loadInfoschema for snapshot infoschema should not be added to schema history.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

It requires restart of tidb, follow the steps in the issue.

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/domain/domain.go [hunk 1]
```diff
@@ -371,7 +373,17 @@ func (do *Domain) loadInfoSchema(startTS uint64, isSnapshot bool) (infoschema.In
...
 	infoschema_metrics.LoadSchemaDurationLoadAll.Observe(time.Since(startTime).Seconds())
 
-	builder := infoschema.NewBuilder(do, do.sysFacHack, do.infoCache.Data, useV2)
+	data := do.infoCache.Data
+	if isSnapshot {
+		// Use a NewData() to avoid adding the snapshot schema to the infoschema history.
+		// Why? imagine that the current schema version is [103 104 105 ...]
+		// Then a snapshot read require infoschem version 53, and it's added
+		// Now the history becomes [53,  ... 103, 104, 105 ...]
+		// Then if a query ask for version 74, we'll mistakenly use 53!
```
Comment (lance6716) [inline pkg/domain/domain.go:382]: And now the query for 74 will fail? Is this expected?

File: pkg/infoschema/test/infoschemav2test/v2_test.go [hunk 2]
```diff
@@ -533,3 +533,39 @@ func TestSnapshotInfoschemaReader(t *testing.T) {
 	sql = fmt.Sprintf("select * from INFORMATION_SCHEMA.TABLES as of timestamp '%s' where table_schema = 'issue55827'", timeStr)
 	tk.MustQuery(sql).Check(testkit.Rows())
 }
+
+
+func TestIssue55835(t *testing.T) {
```
Comment (lance6716) [inline pkg/infoschema/test/infoschemav2test/v2_test.go:538]: The UT can pass without this fix

File: pkg/domain/domain.go (hunk 1 above)
Comment (tiancaiamao) [reply pkg/domain/domain.go:382]: Of course not.
Not we get two kinds of  infoschema, with non-snapshot one [ ... 103, 104, 105 ...]
and snapshot infoschema contains no history like [53]    [74] etc

File: pkg/infoschema/test/infoschemav2test/v2_test.go (hunk 2 above)
Comment (tiancaiamao) [reply pkg/infoschema/test/infoschemav2test/v2_test.go:538]: The problem is, when we switch v1/v2 back and forth, the *history* is not cleared.
So v1/v2 switch can't reproduce the tidb restart scenario.
During adding this test, I did find something wrong: when modify @@tidb_schema_cache_size, without a new DDL operation, v1/v2 switch does not happen, that's not expected.

I confirmed this change can fix the bug and tested manually following the steps described in the issue.

**Comment by tiancaiamao** [comment]
PTAL @lance6716 @wjhuang2016

**Pull request merged**